### ISSUE TITLE
[bug] Incompleteness when matching with `$$$`

### ISSUE THREAD

**Issue by zgrannan** [ast-grep/ast-grep#714]
# ⏯ Playground Link

[Playground link with relevant code](https://ast-grep.github.io/playground.html)

# 💻 Code

Code:

```rust
fn main(){
    if(y) {
        
    }
    if(x) {

    }
    if(x) {
        
    }
}
```

Pattern:

```yml
{ $$$E if($A){$$$B} if($A){$$$C} $$$F}
```

# 🙁 Actual behavior

Does not match. However, it does match if the statement `if(y){}` is removed. 

Surprisingly, the behaviour is correct if one replaces `if(y){}`, with, for example `println!("baz")`. 

# 🙂 Expected behavior

Should match with `$$$E -> if(y){}`, `$A -> x` 

I haven't looked at the code, but I suspect what could be happening is that the metavar `$A` is assigned to `y` when encountering the first `if` statement, but the decision is not backtracked.

**Comment by HerringtonDarkholme** [comment]
Hi, thanks for giving ast-grep a try.

One important thing is that: [Pattern code must be valid code that tree-sitter can parse.](https://ast-grep.github.io/guide/pattern-syntax.html#pattern-matching)

`{ $$$E if($A){$$$B} if($A){$$$C} $$$F}` results a parsing error and it may or may not match the code.

<img width="571" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/92c4684b-acf2-4006-8f10-65c6f084b6b2">


I would suggest to use YAML rule to search the code instead.

**Comment by zgrannan** [comment]
Ah, thanks, I didn't realize that the pattern did not parse! However, this updated version (that parses correctly), has the same issue: [Playground Link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicnVzdCIsInF1ZXJ5IjoiZm4gbWFpbigpIHsgJCQkRTsgaWYoJEEpeyQkJEJ9IGlmKCRBKXskJCRDfSAkJCRGfSIsInJld3JpdGUiOiJmbiBtYWluKCkgeyAkJCRFOyBpZigkQSl7JCQkQiAkJCRDfSAkJCRGfSIsImNvbmZpZyI6IiMgWUFNTCBSdWxlIGlzIG1vcmUgcG93ZXJmdWwhXG4jIGh0dHBzOi8vYXN0LWdyZXAuZ2l0aHViLmlvL2d1aWRlL3J1bGUtY29uZmlnLmh0bWwjcnVsZVxucnVsZTpcbiAgcGF0dGVybjogJEFcbiAga2luZDogY2FsbF9leHByZXNzaW9uXG4gIGluc2lkZTpcbiAgICBraW5kOiBmaWVsZF9pbml0aWFsaXplclxuICAgIGhhczpcbiAgICAgIGtpbmQ6IGZpZWxkX2lkZW50aWZpZXJcbiAgICAgIHJlZ2V4OiBcInNuYXBzaG90X2NvbnN0cnVjdG9yXCJcbmZpeDogU29tZSgkQSkiLCJzb3VyY2UiOiJmbiBtYWluKCl7XG4gICAgXG4gICAgaWYoeSl7XG5cbiAgICB9XG4gICAgcHJpbnRsbiEoXCJiYXpcIik7XG4gICAgXG4gICAgaWYoeCkge1xuXG4gICAgfVxuICAgIGlmKHgpIHtcbiAgICAgICAgXG4gICAgfVxufSJ9)

In particular, for the following code:

```rust
fn main(){
    
    if(y){

    }
    println!("baz");
    
    if(x) {

    }
    if(x) {
        
    }
}
```

Commenting out the `if(y){}` allows the pattern `{ $$$E; if($A){$$$B} if($A){$$$C} $$$F}` to match. But it should match in the original code as well.

> I would suggest using [the YAML rule](https://ast-grep.github.io/guide/rule-config/atomic-rule.html) to search the code instead.

The naive translation into yaml (below) has the same problem:
```
rule:
  pattern: '{ $$$E; if($A){$$$B} if($A){$$$C} $$$F}'
```

At a high-level, what I want to accomplish is to optimise some auto-generated code by removing these redundant `if` statements. I.e translating `if($A){ $$$B} if ($A) {$$$C}` to `if($A) {$$$B $$$C}`. Perhaps there is an easier way to do that than the way I am trying?

**Comment by HerringtonDarkholme** [comment]
The problem of `$$$E;` is that it has a semicolon behind the metavars.  The pattern can only match code that has exactly the same CST structure, that is, has a semicolon at the end of the statement.  See [this link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicnVzdCIsInF1ZXJ5IjoiZm4gbWFpbigpIHsgXG4gICAgJCQkO1xuICAgIGlmKCRBKXskJCRCfSBcbiAgICBpZigkQSl7JCQkQ30gXG4gICAgJCQkRlxufSIsInJld3JpdGUiOiJmbiBtYWluKCkgeyAkJCRFOyBpZigkQSl7JCQkQiAkJCRDfSAkJCRGfSIsImNvbmZpZyI6InJ1bGU6XG4gIGtpbmQ6IGlmX2V4cHJlc3Npb24iLCJzb3VyY2UiOiJmbiBtYWluKCl7XG4gICAgYTsgICAgXG4gICAgaWYoeCkge1xuXG4gICAgfVxuICAgIGlmKHgpIHtcbiAgICAgICAgXG4gICAgfVxufSJ9)

This is an advanced topic in ast-grep. You can see the [ast-grep's core concept deep dive](https://ast-grep.github.io/advanced/core-concepts.html#significant-vs-trivial) here for more context. It is generally hard to parse variations of meta-variable codes.

However, using YAML rule is pretty easy for this case. The change is basically merging two if statements to one, if I understand it correct.

So it can be translated to:

> One [if statement](https://ast-grep.github.io/guide/rule-config/atomic-rule.html#kind) should [follow](https://ast-grep.github.io/reference/rule.html#follows) the other if-statement, [skipping](https://ast-grep.github.io/guide/rule-config/relational-rule.html#fine-tuning-relational-rule) all AST nodes between. 


```YAML
rule:
  pattern: if ($A) { $$$B }
  follows:
    pattern: if ($A) { $$$C }
    stopBy: end
fix: |-
  if ($A) { 
    $$$C
    $$$B
  }
```

And we need another rule to delete duplicate if statements. 

```yaml
rule:
  pattern: if ($A) { $$$B }
  precedes:
    pattern: if ($A) { $$$C }
    stopBy: end
fix: ''
```


We can utilize ast-grep's [multiple rules in the same file](https://ast-grep.github.io/reference/playground.html#test-multiple-rules) feature.



The [final result](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InJ1c3QiLCJxdWVyeSI6ImZuIG1haW4oKSB7IFxuICAgICQkJDtcbiAgICBpZigkQSl7JCQkQn0gXG4gICAgaWYoJEEpeyQkJEN9IFxuICAgICQkJEZcbn0iLCJyZXdyaXRlIjoiZm4gbWFpbigpIHsgJCQkRTsgaWYoJEEpeyQkJEIgJCQkQ30gJCQkRn0iLCJjb25maWciOiJydWxlOlxuICBwYXR0ZXJuOiBpZiAoJEEpIHsgJCQkQiB9XG4gIGZvbGxvd3M6XG4gICAgcGF0dGVybjogaWYgKCRBKSB7ICQkJEMgfVxuICAgIHN0b3BCeTogZW5kXG5maXg6IHwtXG4gIGlmICgkQSkgeyBcbiAgICAkJCRDXG4gICAgJCQkQlxuICB9XG4tLS1cbnJ1bGU6XG4gIHBhdHRlcm46IGlmICgkQSkgeyAkJCRCIH1cbiAgcHJlY2VkZXM6XG4gICAgcGF0dGVybjogaWYgKCRBKSB7ICQkJEMgfVxuICAgIHN0b3BCeTogZW5kXG5maXg6ICcnIiwic291cmNlIjoiZm4gbWFpbigpe1xuICAgIGlmKHkpIHtcbiAgICAgICAgXG4gICAgfVxuICAgIGlmKHgpIHtcbiAgICAgICAgcHJpbnRsbiEoXCJIZXJlIVwiKTtcbiAgICB9XG4gICAgaWYoeCkge1xuICAgICAgICBwcmludGxuIShcIlRoZXJlIVwiKTsgXG4gICAgfVxufSJ9)

<img width="1013" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/6144a3b8-0df9-4ae7-ae54-8e6f842396d9">

**Comment by HerringtonDarkholme** [comment]
It is really interesting to see how multiple meta var works in this example.

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicnVzdCIsInF1ZXJ5IjoiZm4gbWFpbigpIHsgXG4gICAgJCQkUFJFXG4gICAgaWYoJEEpeyQkJEJ9IFxuICAgICQkJE1JRFxuICAgIGlmKCRBKXskJCRDfSBcbiAgICAkJCRBRlRFUlxufSIsInJld3JpdGUiOiJmbiBtYWluKCkgeyBcbiAgICAkJCRQUkVcbiAgICAkJCRNSURcbiAgICBpZigkQSl7XG4gICAgICAgICQkJEMgXG4gICAgICAgICQkJEJcbiAgICB9IFxuICAgICQkJEFGVEVSXG59IiwiY29uZmlnIjoicnVsZTpcbiAgcGF0dGVybjogaWYgKCRBKSB7ICQkJEIgfVxuICBmb2xsb3dzOlxuICAgIHBhdHRlcm46IGlmICgkQSkgeyAkJCRDIH1cbiAgICBzdG9wQnk6IGVuZFxuZml4OiB8LVxuICBpZiAoJEEpIHsgXG4gICAgJCQkQ1xuICAgICQkJEJcbiAgfVxuLS0tXG5ydWxlOlxuICBwYXR0ZXJuOiBpZiAoJEEpIHsgJCQkQiB9XG4gIHByZWNlZGVzOlxuICAgIHBhdHRlcm46IGlmICgkQSkgeyAkJCRDIH1cbiAgICBzdG9wQnk6IGVuZFxuZml4OiAnJyIsInNvdXJjZSI6ImZuIG1haW4oKXtcbiAgICB3aGlsZSAodGVzdCkge1xuICAgIH1cbiAgICAvLyBpZiAoeSkge31cbiAgICBpZih4KSB7XG4gICAgICAgIHByaW50bG4hKFwiSGVyZSFcIik7XG4gICAgfVxuICAgIHNvbWVfc3RhdGVtZW50cygpO1xuICAgIC8vIGlmICh5KSB7fVxuICAgIHdlYl8yKCk7XG4gICAgaWYoeCkge1xuICAgICAgICBwcmludGxuIShcIlRoZXJlIVwiKTsgXG4gICAgfVxufSJ9

**Comment by zgrannan** [comment]
Thank you for the reply! It's nice to see that this is possible to achieve and learn more about `ast-grep`

I don't suppose it's possible to match the entire scope of the relational rule (i.e. including the related part) as the source for the `fix`, so that this could be done in one rule?

In particular, I'm worried that the multi-rule approach could be fragile in practice. The particular example is:

```rust
if(x){
  if(y) { print!("1") } 
}
if(x){
  if(y) { print!("2") } 
}
```

after the first rule should rewrite to:

```rust
if(x){
  if(y) { print!("1"); } 
}
if(x){
  if(y) { print!("1"); } 
  if(y) { print!("2"); } 
}
```

However, now the "deletion" rule also should in theory delete not only the:

```rust
if(x){
  if(y) { print!("1"); } 
}
```

But also in the second `if(x)` branch, it presumably deletes:
```rust
  if(y) { print!("1"); } 
```
Resulting in the invalid ultimate transformation to:
```rust
if(x){
  if(y) { print!("2"); } 
}
```

Actually, I couldn't replicate this behaviour in the playground (not sure why). But it seems like that's what would be expected.

**Comment by HerringtonDarkholme** [comment]
Hey @zgrannan ! I am glad you discovered the hard part for fixing! 😄 

I have encountered issues about nested matches and replacement before(in the form of panic).
And it is unclear about the most reasonable behavior about how to orchestrate the replacement of conflicting rules.

So I took the easiest approach: skip all replacement that overlaps preceding fixes.

So in this case,

**Comment by zgrannan** [comment]
Okay, that makes sense, thanks! I don't think there's anything that really needs to be done about this issue for now, so I'll close it.
